data-structures/hashmap.md

```diff
@@ -295,3 +298,55 @@ The impact on the data structure layout imposed by the use of buckets in the IPL
 ### Security
 
 As yet, there is no known hash collision attack vector against IPLD data structures. There may conceivably be use cases where user-input is able to impact `key`s and collisions against a chosen `hashAlg` are practical. In such cases, an IPLD HashMap could be built whereby it reaches its maximum depth of `(digestLength x 8) / bitWidth` quickly and further colliding additions cause errors and possible denial of service. The current use-cases of IPLD do not lend themselves to denial of service attacks of this kind. Further practical application and research may change this understanding and dictate the need for hash algorithms with large byte output and/or cryptographic hash algorithms without known collision flaws.
+
+## Appendix: Filecoin HAMT Variant
+
+The [Filecoin Project blockchain](https://filecoin-project.github.io/specs/) makes use of a HAMT that uses the same HAMT with buckets and CHAMP mutation semantics as outlined in this document. It encodes directly as [DAG-CBOR](../block-layer/codecs/dag-cbor.md) but uses a different block layout to the one specified here. This section documents the specific ways that the Filecoin HAMT variant differs from this specification. IPLD HashMap implementations may be able to implement a form that provides compatibility with Filecoin when requested by the user.
+
+The reference Go implementation for the Filecoin HAMT is used by the [Lotus](https://lotu.sh/) client and is available at <https://github.com/ipfs/go-hamt-ipld>.
+
+### Implicit and fixed parameters
+
+The Filecoin HAMT _does not_ use an explicit root block (`HashMapRoot`) to encode its parameters within the data. Instead it is expected that consumers of the data understand the parameters from a combination of the Filecoin specification and versioning of the blockchain over time. All HAMT nodes take the same form, there is no differentiation for a root node and an implementation must bring implicit parameters when decoding each node.
+
+* `hashAlg`: The hash algorithm used by the Filecoin HAMT is the x64 form of the 64-bit [MurmurHash3](https://github.com/aappleby/smhasher). _(TODO: this may be changed to SHA2-256 prior to Filecoin mainnet: https://github.com/filecoin-project/specs-actors/issues/517)_
+* `bitWidth`: The Filecoin HAMT fixes the bit width to `5`, meaning that each node of the HAMT can contain up to `2`<sup>`5`</sup> (`32`) elements containing either buckets or links to child nodes.
+* `bucketSize`: The Filecoin HAMT fixes the maximum length of its buckets to `3`, meaning a maximally full HAMT leaf node can contain `32 x 3` (`96`) key/value pairs.
```

Comment:
yeah each node is capped at 3 statically, might want to reword these. the bit width is potential values to avoid collisions at each level

Reply:
Sorry, I'm not sure what you're suggesting here, can you suggest some text changes that might make it clearer in your view? This text is also building on a lot of content that already exists in this document that explains the algorithm and how these things all work together so it's assuming you're using all of that as a reference and these details are just the Filecoin HAMT specific bits.

Reply:
also, in https://github.com/ipfs/go-hamt-ipld/pull/52 I've proposed renaming `arrayWidth` to align with the terminology here to save confusion, I figure that it's easier to do it in the Go code than in this spec and all the JavaScript code that implements this spec